src/json/def_to_json.cr

```diff
@@ -0,0 +1,176 @@
+      json.object do
+        {% for key, options in mappings %}
+          {% unless options.is_a?(HashLiteral) || options.is_a?(NamedTupleLiteral) %}
+            {% options = {nil: nil} %}
+          {% end %}
+          ::JSON.field_to_json({{(options[:property] || key).id}}, {{key.id.stringify}}, {{options}})
+        {% end %}
+      end
+    end
+  end
+
+  # This is a convenience method to allow invoking `JSON.def_to_json`
+  # with named arguments instead of with a hash/named-tuple literal.
+  macro def_to_json(**mappings)
+    ::JSON.def_to_json({{mappings}})
+  end
+
+  # This `def_to_json` macro generates a `to_json(value : {{type}}, json : JSON::Builder)` method based on provided type and mappings.
+  #
+  # This is useful to create a JSON converter which responds to a number of `to_json(value, builder : JSON::Builder)` overloads.
+  #
+  # In contrast to the `to_json(JSON::Builder)` method defined by `def_to_json(mapping)` the values for this JSON generator
+  # are retrieved from calling a method on the object provided as the first argument whose type is declared as first argument
+  # to this macro.
+  #
+  # ```
+  # require "json"
+  #
+  # module NeighborhoodConverter
+  #   extend self
+  #   JSON.def_to_json(House, {street_number: true})
+  # end
+  #
+  # class House
+  #   getter street : String
+  #   getter street_number : Int32
+  #   getter neighbor : House? = nil
+  #
+  #   JSON.def_to_json(
+  #     address: true,
+  #     neighbor: {converter: NeighborhoodConverter}
+  #   )
+  #
+  #   def initialize(@street, @street_number, @neighbor = nil)
+  #   end
+  #
+  #   def address
+  #     "#{street} #{street_number}"
+  #   end
+  # end
+  #
+  # neighbor = House.new("Crystal Road", 1235)
+  # house = House.new("Crystal Road", 1234, neighbor)
+  # house.to_json # => %({"address":"Crystal Road 1234","neighbor":{"street_number":1235}})
+  # ```
+  macro def_to_json(type, mappings)
+    def to_json(value : {{type.id}}, json : ::JSON::Builder)
+      json.object do
+        {% for key, options in mappings %}
+          {% unless options.is_a?(HashLiteral) || options.is_a?(NamedTupleLiteral) %}
+            {% options = {nil: nil} %}
+          {% end %}
+          ::JSON.field_to_json(value.{{(options[:property] || key).id}}, {{key.id.stringify}}, {{options}})
+        {% end %}
+      end
+    end
+  end
+
+  # This is a convenience method to allow invoking `JSON.def_to_json`
+  # with named arguments instead of with a hash/named-tuple literal.
+  macro def_to_json(type, **mappings)
+    ::JSON.def_to_json({{type}}, {{mappings}})
+  end
+
+  # The `StringConverter` has a class method `to_json` wich can be used as a converter for `JSON.def_to_json`. The value is added
+  # to the builder as a string.
+  module StringConverter
+    def self.to_json(value, builder)
+      value.to_s.to_json(builder)
+    end
+  end
+
+  # :nodoc:
+  macro field_to_json(value_name, field_name, options)
```

Comment:
I guess I did this a bit sloppy because it is only an internal macro and was refactored quite a few times... ;)

But I like your suggestions, thanks for reviewing!
* I think I'll change the macro name to `emit_value_to_json` - it's better to have a verb.
* I am not sure about `field_name` because it is technically not always referring to a field but might as well be an ordinary method returning a value unrelated to an instance variable.